UI: Enable multiple audio tracks in Simple Output recording

pkviet opened the pull request:
### Description
This adds support for multiple audio tracks in Simple Output for recordings. This is enabled for all quality presets (including "Lossless") except "Same as Stream".
 This is also enabled for the Replay Buffer.
An exception is made for flv recording format since it only allows a single audio track.
The recorded track (and streaming track) is then Track 1 as before.

**Screenshots**
lossless preset:
![obs64_2022-07-23_20-08-47](https://user-images.githubusercontent.com/9283407/180618839-68421c4d-c723-4654-9814-ed013fcf9716.png)

encoded presets:
![obs64_2022-07-23_20-08-27](https://user-images.githubusercontent.com/9283407/180618858-b105d8e2-c130-407c-936f-1f5852bda47f.png)

flv:
![obs64_2023-02-27_12-51-12](https://user-images.githubusercontent.com/9283407/221556752-2497071c-a64d-437e-977b-ecdc6b639cc9.png)

same as stream preset:
![obs64_2023-02-27_12-51-35](https://user-images.githubusercontent.com/9283407/221556829-4b94ca47-5eab-47ec-9cc1-56fb9dc68dbf.png)

### Motivation and Context
Members of obs support team report that regularly people switch from Simple to Advanced output just to be able to record multi track audio.
It would be better though for many to stick to Simple output, since it is ... simpler.

### How Has This Been Tested?
Recordings made with three tracks in Simple output for:
- lossless preset,
- encoded presets (except flv).
Checked that flv & "same as stream" preset record track 1.

### Types of changes
 - New feature (non-breaking change which adds functionality) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

WizardCM commented:
Note to self: test AAP.

Fenrirthviti commented:
From internal discussions about this: Disappearing items is something we no longer want to present in OBS, so this will need to find a permanent location to live. Disabling the checkboxes might be a solution here, but we will probably need to have a larger discussion on how best to approach this from a UI/UX standpoint for the various options.

Having this PR open should push those conversations forward as a high priority (at least, for myself) for post-v28 features.

CrxckedDest0y3r reviewed (commented):
[deleted]

Warchamp7 commented:
Can we get a radio group selection for flv (and any other formats that don't support multiple audio tracks) instead of removing the ?

If we are hardcoded to track 1 for single track formats, I'd still like a radio group and simply have it disabled.

pkviet commented:
Updated the PR to implement @Fenrirthviti & @Warchamp7 suggestions.
When a preset / container which is not compatible with multiple tracks is picked, radio buttons are greyed out (track 1 is displayed as selected since it's the default for Simple Output; capability to pick the track for flv is beyond the purview of this PR.)
See the updated screen captures in first post of the PR.

RytoEX commented on UI/window-basic-main.cpp:
```diff
@@ -1507,7 +1507,8 @@ bool OBSBasic::InitBasicConfigDefaults()
 	config_set_default_string(basicConfig, "SimpleOutput",
 				  "RecAudioEncoder", "aac");
-
+	config_set_default_uint(basicConfig, "SimpleOutput", "SimpleRecTracks",
+				(1 << 0));
```
The blank space here visually separated Simple Output and Advanced Output settings. Re-add it below the new Simple Output setting to keep them separated.

pkviet replied:
done

RytoEX commented on UI/window-basic-settings.cpp:
```diff
@@ -497,6 +497,12 @@ OBSBasicSettings::OBSBasicSettings(QWidget *parent)
 	HookWidget(ui->simpleReplayBuf,      CHECK_CHANGED,  OUTPUTS_CHANGED);
 	HookWidget(ui->simpleRBSecMax,       SCROLL_CHANGED, OUTPUTS_CHANGED);
 	HookWidget(ui->simpleRBMegsMax,      SCROLL_CHANGED, OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack1,   CHECK_CHANGED,  OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack2,   CHECK_CHANGED,  OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack3,   CHECK_CHANGED,  OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack4,   CHECK_CHANGED,  OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack5,   CHECK_CHANGED,  OUTPUTS_CHANGED);
+	HookWidget(ui->simpleOutRecTrack6,   CHECK_CHANGED,  OUTPUTS_CHANGED);
```
These should probably be grouped with the other Simple Output Mode (`simpleOut`) settings above, in an appropriate location relative to the other controls, if possible.

pkviet replied:
done

RytoEX commented on UI/window-basic-settings.cpp:
```diff
@@ -5267,12 +5290,30 @@ void OBSBasicSettings::SimpleReplayBufferChanged()
 	bool replayBufferEnabled = ui->simpleReplayBuf->isChecked();
 	bool lossless = qual == "Lossless";
 	bool streamQuality = qual == "Stream";
-
+	int abitrate = 0;
```
Re-add the blank line below this newly added line.

pkviet replied:
done

tytan652 commented on UI/window-basic-main.cpp:
```diff
@@ -1507,6 +1507,8 @@ bool OBSBasic::InitBasicConfigDefaults()
 				  "StreamAudioEncoder", "aac");
 	config_set_default_string(basicConfig, "SimpleOutput",
 				  "RecAudioEncoder", "aac");
+	config_set_default_uint(basicConfig, "SimpleOutput", "SimpleRecTracks",
```
It's already in the "SimpleOutput", having "Simple" in the option name is quiet redundant.

pkviet replied:
true

pkviet replied:
done

tytan652 commented on UI/window-basic-settings.cpp:
```diff
@@ -1925,6 +1931,15 @@ void OBSBasicSettings::LoadSimpleOutputSettings()
 		config_get_int(main->Config(), "SimpleOutput", "RecRBTime");
 	int rbSize =
 		config_get_int(main->Config(), "SimpleOutput", "RecRBSize");
+	int tracks = config_get_int(main->Config(), "SimpleOutput",
+				    "SimpleRecTracks");
```
Same

pkviet replied:
done

pkviet commented:
**Update**

- Per request from @Fenrirthviti and @Warchamp7 , moved the track boxes below the audio encoder field.
![obs64_2023-03-23_20-10-10](https://user-images.githubusercontent.com/9283407/227324673-2fbce1b9-fbcc-4dcf-9e65-b6960d003c90.png)

- There is an incompatibility between mov container and opus codec, introduced by the addition of opus to the codecs in simple output recording in https://github.com/obsproject/obs-studio/pull/8090 ; this will be fixed later through https://github.com/obsproject/obs-studio/pull/8498

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1193,13 +1255,35 @@ void SimpleOutput::UpdateRecording()
+			for (int i = 0; i < MAX_AUDIO_MIXES; i++) {
+				if ((tracks & (1 << i)) != 0) {
+					obs_output_set_audio_encoder(
+						fileOutput, audioTrack[i], idx);
+					idx++;
```
Can `idx++` not be here?
```
obs_output_set_audio_encoder(fileOutput, audioTrack[i], idx++);
```

pkviet replied:
sure

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1193,13 +1255,35 @@ void SimpleOutput::UpdateRecording()
+		} else {
+			for (int i = 0; i < MAX_AUDIO_MIXES; i++) {
+				if ((tracks & (1 << i)) != 0) {
+					obs_output_set_audio_encoder(
+						replayBuffer, audioTrack[i],
+						idx2);
+					idx2++;
```
Can `idx2++` not be here?
```
obs_output_set_audio_encoder(replayBuffer, audioTrack[i], idx2++);
```

pkviet replied:
done

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1193,13 +1255,35 @@ void SimpleOutput::UpdateRecording()
+			}
+		}
 	}
-
```
Restore this blank line.

pkviet replied:
done

pkviet commented on UI/window-basic-settings.cpp:
```diff
@@ -3770,6 +3785,14 @@ void OBSBasicSettings::SaveOutputSettings()
 	SaveSpinBox(ui->simpleRBSecMax, "SimpleOutput", "RecRBTime");
 	SaveSpinBox(ui->simpleRBMegsMax, "SimpleOutput", "RecRBSize");
+	config_set_int(
+		main->Config(), "SimpleOutput", "SimpleRecTracks",
```
done

RytoEX reviewed (approved):
Looks okay at a glance.

Warchamp7 commented:
Approved on my end

pkviet commented:
**Update**
- discovered today that this Qt bug is fixed: https://bugreports.qt.io/browse/QTBUG-6864
- this allows now the use of a QtStackedWidgets, which makes for less code and cleaner. So I've switched to that.
- also added one commit to move the tracks in Advanced Output > Standard Recording per request of @Warchamp7 .
- also moved the video encoder field above that of audio encoder  in Advanced Output > Standard Recording per request of @derrod 
![obs64_2023-03-25_21-30-19](https://user-images.githubusercontent.com/9283407/227740151-7312a3a9-b8e9-44be-aba2-c1db6366bf9d.png)

RytoEX reviewed (approved):
Looks fine at a glance.
